source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,35 @@
 class UrlsController < ApplicationController
+
+  def index
+    @urls = Url.all
+  end
+
+  def show
+    if params[:id].start_with?("!")
```

Comment:
Let's try to make the router send these two kinds of requests to two controller actions. Let me know if you'd like some guidance on that.